new BIP: codex32

apoelstra opened the pull request:
This introduces "codex32", a wallet seed/share format that uses the bech32 alphabet, a bech32-like checksum, and supports Shamir Secret Sharing, all possible without the use of electronic computers (although you can, and for some steps probably should, use computers if you want).

Mailing list discussion: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-February/021469.html
Website with more docs about hand computation (which isn't really covered by the BIP): https://secretcodex32.com/index.html

apoelstra committed:
new BIP: codex32

roconnor-blockstream commented:
Add a comment that notes that if you are resharing a secret by taking existing shares an recovering the secret, tossing out the old shares and making new ones, it is important to create a fresh unique ID.  Without a fresh ID you run the risk of accidentally combining old shares with new shares (due to forgetting to destroy some old shared), which could lead to confusion and/or the apparent loss of funds.

roconnor-blockstream commented:
After discussion with Christopher Allen et. al., I've opened PR https://github.com/apoelstra/bips/pull/2 for adding a rationale section for not encoding BIP-0039 entropy.

I suppose comments on that proposed amendment should be directed to that PR, while this PR remains open for assigning a BIP number and other comments on this proposal.

roconnor-blockstream commented:
Recommend noting master xprvs in each test case.

luke-jr commented on bip-0000.mediawiki:
```diff
@@ -0,0 +1,472 @@
+<pre>
+  BIP: ????
+  Layer: Applications
+  Title: codex32
```
Would prefer a title that means something

roconnor-blockstream replied:
Yes.

Maybe something like `Codex32 format for Shamir Secret Shares of [BIP-0032] master seeds`.

apoelstra replied:
Sure. How about "codex32: a bech32-based seed encoding supporting Shamir Secret Sharing and hand computation"?

roconnor-blockstream replied:
SGTM.

apoelstra replied:
Pushed.

michaelfolkson commented:
It'd be nice to get SLIP39 BIPed as well as this at some point in the future (especially as this draft BIP references it) if that factors into the BIP number allocation and possible number grouping(s).

vicariousdrama commented:
NACK - Not in keeping with the Bitcoin Philosophy

luke-jr commented:
I stumbled across a memo from 93 BC that appears to say this has been assigned BIP number 93.

apoelstra committed:
assign BIP #93; set type to "Informational"

apoelstra commented:
Thanks @luke-jr!

I have squashed all the existing commits and added a new one which sets the number (and sets the type to "Informational", which matches my intuition about this and also matches BIP 173, which is similar in spirit).

We are not quite ready to merge -- we have a new reference implementation we want to link to (with Postscript and Rust, for now) and we want to replace some test vectors with nicer ones.

apoelstra committed:
change title to fit under length limit

roconnor-blockstream committed:
Updated set of invalid test vectors

Using the identifier "faux" for invalid test vectors.

Extending the set of example a little bit.

kallewoof commented:
Just a reminder to add the BIP entry to `README.mediawiki` as well.

apoelstra committed:
add links to reference impls

apoelstra committed:
add bip93 to README

roconnor-blockstream reviewed (approved):
LGTM

apoelstra commented:
@kallewoof I think we are ready to merge this

murchandamus commented:
@roconnor-blockstream, @apoelstra: I was looking through the BIPs in Draft status and was wondering whether you have considered advancing BIP93 to Complete, perhaps after the open PRs by Ben are resolved.
